### PR TITLE
ability to filter badges based on state

### PR DESCRIPTION
**Pull request by iantrich**
Closes home-assistant/ui-schema#165

```yaml
    badges:
      - sensor.yr_symbol
      - binary_sensor.movement_backyard
      - person.ian
      - entity: sensor.outside_humidity
        state_filter:
          - operator: '>'
            value: 55
      - binary_sensor.basement_floor_wet
      - input_number.slider2
```

Docs: https://github.com/home-assistant/home-assistant.io/pull/10499

### CODE REVIEW

File: src/panels/lovelace/editor/view-editor/hui-edit-view.ts [hunk 1]
```diff
@@ -216,7 +216,9 @@ export class HuiEditView extends LitElement {
 
     const viewConf: LovelaceViewConfig = {
       cards: this._cards,
-      badges: this._badges!.map((entityConf) => entityConf.entity),
+      badges: this._badges!.map((entityConf) =>
+        Object.keys(entityConf).length > 1 ? entityConf : entityConf.entity
```
Comment (bramkragten) [inline src/panels/lovelace/editor/view-editor/hui-edit-view.ts]: We can save all new ones/edits as objects, as long as we keep loading strings to stay backwards compatible.

File: src/panels/lovelace/views/hui-view.ts [hunk 2]
```diff
@@ -263,12 +263,38 @@ export class HUIView extends LitElement {
 
     const elements: HUIView["_badges"] = [];
     // It's possible that a null value was stored as a badge entry
-    const badges = processConfigEntities(config.badges.filter(Boolean));
+    const badges = processConfigEntities(
+      config.badges.filter((badgeConf) => {
+        if (typeof badgeConf === "object") {
```
Comment (bramkragten) [inline src/panels/lovelace/views/hui-view.ts]:
```suggestion
        if (badgeConf && typeof badgeConf !== "object") {
            return true;
        }
```

File: src/panels/lovelace/views/hui-view.ts [hunk 3]
```diff
@@ -263,12 +263,38 @@ export class HUIView extends LitElement {
...
+      })
+    );
+
     for (const badge of badges) {
-      const element = document.createElement("ha-state-label-badge");
       const entityId = badge.entity;
+      const stateObj = this.hass!.states[entityId];
+      const element = document.createElement("ha-state-label-badge");
       element.hass = this.hass;
-      element.state = this.hass!.states[entityId];
```
Comment (bramkragten) [inline src/panels/lovelace/views/hui-view.ts]: Why was this changed?

Comment (iantrich) [reply src/panels/lovelace/views/hui-view.ts]: Some things moved around while testing, shouldn't have checked this part in :+1:

File: src/panels/lovelace/editor/view-editor/hui-edit-view.ts (hunk 1 above)
Comment (iantrich) [reply src/panels/lovelace/editor/view-editor/hui-edit-view.ts]: is one really preferable over the other?

Comment (bramkragten) [reply src/panels/lovelace/editor/view-editor/hui-edit-view.ts]: I saves us from doing this check, keeps things simpler

**Comment by balloob** [comment]
I thought the idea would be that badge config is going to be like cards, and we add a filter badge. https://github.com/home-assistant/home-assistant-polymer/pull/3600#issuecomment-528000430

```yaml
- type: filter
   filter: …
   config:
     entity: binary_sensor.bla
```

**Comment by iantrich** [comment]
@balloob I went threw an iteration like that but didn't like the clunkiness of it and saw no real reason to support additional badge types. This, to me, is much cleaner.

**Comment by balloob** [comment]
Isn't it the same config set up as we have for cards ? 

I think that by conflating concerns, we will block the path to allow new types, like 'activate scene' or custom ones to be added as badges.

**Comment by iantrich** [comment]
Yeah, I had it written up and then 2nd guessed adding the complexity as I didn't really see a need for different badge types or custom ones 🤷‍♀️

**Comment by iantrich** [comment]
will be back with yet another PR for this later ;)
